### ISSUE TITLE
Trash bin auto deletion broken

### ISSUE THREAD

**Issue by steve-s-71** [owncloud/core#22537]
### Steps to reproduce
1. Delete file
2. Wait 7 Months
3. Still find it in Trash bin

### Expected behaviour
Trash bin auto deletion should delete files after 30 days (default) or what is configured in config.php

### Actual behaviour
There are many months old files in the Trash bin. If not manually deleted from the Trash bin, it seems they are hold for ever. As it stands, all user accounts are impacted by this issue.

I tried following in the config file with no success:
'trashbin_retention_obligation' => 14,

also I tried:
'trashbin_retention_obligation' => '14',

also I tried:
'trashbin_retention_obligation' => 14,
'trashbin_auto_expire' => true,

also I tried:
'trashbin_retention_obligation' => 7, 14,

finally I tried:
'trashbin_retention_obligation' => '7, 14',

all with no success.

### Server configuration
**Operating system**:
Linux

**Web server:**
Apache

**Database:**
MySQL

**PHP version:**
5.?

**ownCloud version:** (see ownCloud admin page)
8.2.2

**Updated from an older ownCloud or fresh install:**
updated. originally it was 7.x or 6.x I dont remember

**Comment by nickvergessen** [comment]
Please provide us with the config information that is mentioned on the issue template:

> **The content of config/config.php:**
> 
> ```
If you have access to your command line run e.g.:
sudo -u www-data php occ config:list system
from within your ownCloud installation folder
> 
> or 
> 
> Insert your config.php content here
(Without the database password, passwordsalt and secret)
```

**Comment by steve-s-71** [comment]
Hi Nick,

Here is current config.php content. Instead of  'trashbin_retention_obligation' => '14, 30',  I also tried the other stuff as listed above in my inital post:

<?php
$CONFIG = array (
  'instanceid' => '#####',
  'passwordsalt' => '#####',
  'secret' => '#####',
  'trusted_domains' =>
  array (
    0 => '192.168.178.3',
    1 => '#####',
    2 => '#####',
  ),
  'datadirectory' => '/ocdata/data',
  'overwrite.cli.url' => '#####/owncloud',
  'dbtype' => 'mysql',
  'version' => '8.2.2.2',
  'dbname' => 'owncloud',
  'dbhost' => 'localhost',
  'dbtableprefix' => 'oc_',
  'dbuser' => '#####',
  'dbpassword' => '#####',
  'installed' => true,
  'theme' => '',
  'maintenance' => false,
  'forcessl' => false,
  'loglevel' => 4,
  'mail_smtpmode' => 'smtp',
  'mail_smtpauth' => 1,
  'mail_from_address' => 'owncloud',
  'mail_domain' => '#####',
  'mail_smtphost' => '#####',
  'mail_smtpport' => '465',
  'mail_smtpname' => '#####',
  'mail_smtppassword' => '#####',
  'mail_smtpsecure' => 'ssl',
  'default_language' => 'de_DE',
  'mail_smtpauthtype' => 'LOGIN',
  'trashbin_retention_obligation' => '14, 30',
);

**Comment by steve-s-71** [comment]
reopen as closing was not intentionally..

**Comment by nickvergessen** [comment]
Okay, seems like you are using ajax cron, since you didn't change it to system or webcron.

Can you have a look at the oc_jobs table in your database?
Would be nice if you can run the following query and post the result here:
```sql
SELECT `class`, COUNT(*) AS num_entries, MIN(`last_run`) AS oldest, MAX(`last_run`) AS newest
FROM `oc_jobs`
GROUP BY `class`
```

**Comment by steve-s-71** [comment]
I have configured background jobs as system cron in the web interface. Under Administrator Tab OC confirms it runs every 15min. 

I am afrait I dont have a tool to run database queries. Can you point me to a simple to install tool?

**Comment by steve-s-71** [comment]
Hi Nick,
I found a way to run the quirey from the bash. Here is the result:

```
+-------------------------------------------------+-------------+------------+------------+
| class                                           | num_entries | oldest     | newest     |
+-------------------------------------------------+-------------+------------+------------+
| OCA\Activity\BackgroundJob\EmailNotification    |           1 | 1456521301 | 1456521301 |
| OCA\Activity\BackgroundJob\ExpireActivities     |           1 | 1456513201 | 1456513201 |
| OCA\Files_Versions\BackgroundJob\ExpireVersions |           1 | 1456521301 | 1456521301 |
| OCA\Files_sharing\Lib\DeleteOrphanedSharesJob   |           1 | 1456521301 | 1456521301 |
| OC\BackgroundJob\Legacy\RegularJob              |           1 | 1456521301 | 1456521301 |
+-------------------------------------------------+-------------+------------+------------+
5 rows in set (0.00 sec)
```

**Comment by smatthie** [comment]
Broken for me too. I have 'trashbin_retention_obligation' => 'auto,30' (which should keep files for a maximum of 30 days, right?) and webcron, but flies up to 8 months old in the trashbin, on OC 8.2.2

**Comment by smatthie** [comment]
When I run "sudo -u www-data php cron.php" in the shell, it removes old files from the trashbin as expected. With webcron, it doesn't. As far as I can see, my webcron is setup correctly: I checked the option in the administrator panel, it shows the last run, and when I the URL in a browser I get the message "success"

**Comment by steve-s-71** [comment]
I ran "sudo -u www-data php cron.php" in the shell, like what smatthie did. Unfortunately it still does not removes old files from the trash bin. Thus same behavior, no matter if I run cron.php from shell, automatically by system setup, Ajax or Webcron (when I configure Webcron and enter URL in a browser it confirm "sucess").

**Comment by nickvergessen** [comment]
Seems like the trashbin background job is missing for you:
```sql
INSERT INTO `oc_jobs`(`class`, `argument`, `last_run`) VALUES ('OCA\Files_Trashbin\BackgroundJob\ExpireTrash', 'null', 0)
```

That should readd the job.
I assume you are suffering from https://github.com/owncloud/core/issues/20399
So applying https://github.com/owncloud/core/pull/21439/files should help

**Comment by steve-s-71** [comment]
I have sucessfully insert the line into oc_jobs and modified lib/private/backgroundjob/job.php accoding to the link (correctly I modified /var/www/owncloud/lib/private/backgroundjob/job.php)

I cant find the other file what shall get modified (tests/lib/backgroundjob/job.php). There is not a folder 'tests' existing as subfolder of  /var/www/owncloud

**Comment by nickvergessen** [comment]
Yeah tests are missing if you are not running a git installation, so that should be fine.

**Comment by steve-s-71** [comment]
Unfortunately the Trash-bin is still not cleaned. I regognised after inserting the line into oc_jobs it looks like this:

```
`
+-------------------------------------------------+-------------+------------+------------+
| class                                           | num_entries | oldest     | newest     |
+-------------------------------------------------+-------------+------------+------------+
| OCAFiles_TrashbinBackgroundJobExpireTrash       |           1 |          0 |          0 |
| OCA\Activity\BackgroundJob\EmailNotification    |           1 | 1456763669 | 1456763669 |
| OCA\Activity\BackgroundJob\ExpireActivities     |           1 | 1456686901 | 1456686901 |
| OCA\Files_Versions\BackgroundJob\ExpireVersions |           1 | 1456763729 | 1456763729 |
| OCA\Files_sharing\Lib\DeleteOrphanedSharesJob   |           1 | 1456763673 | 1456763673 |
| OC\BackgroundJob\Legacy\RegularJob              |           1 | 1456763729 | 1456763729 |
+-------------------------------------------------+-------------+------------+------------+
`
```

then after I ran sudo -u www-data php cron.php the fist line is deleted: 


```
`
+-------------------------------------------------+-------------+------------+------------+
| class                                           | num_entries | oldest     | newest     |
+-------------------------------------------------+-------------+------------+------------+
| OCA\Activity\BackgroundJob\EmailNotification    |           1 | 1456763669 | 1456763669 |
| OCA\Activity\BackgroundJob\ExpireActivities     |           1 | 1456686901 | 1456686901 |
| OCA\Files_Versions\BackgroundJob\ExpireVersions |           1 | 1456763729 | 1456763729 |
| OCA\Files_sharing\Lib\DeleteOrphanedSharesJob   |           1 | 1456763673 | 1456763673 |
| OC\BackgroundJob\Legacy\RegularJob              |           1 | 1456764527 | 1456764527 |
+-------------------------------------------------+-------------+------------+------------+

`
```

I have double checked that I have definitely removed the line from lib/private/backgroundjob/job.php per the link you provided.

**Comment by steve-s-71** [comment]
I further figured out, first line is NOT deleted when the cron.php runs via ajax, webcron or systemcron. First line is  deleted only when I manually enter sudo -u www-data php cron.php into bash

But in any case old files are not removed from the trash bin.

**Comment by nickvergessen** [comment]
Any entry in your log file?

**Comment by steve-s-71** [comment]
In the log there is only 1 continuous Warning: 

"no app in context", "Installation is in transit between the old Encryption (ownCloud <= 8.0) and the new encryption. Please enable the "Default encryption module" and run 'occ encryption:migrate". 

I don't want/need encryption as it cost performance and the sever is located in my own house.

I got the issue fixed that fist line is deleted. Apparently this was caused by missing backslashes in the path what was entred into the database. 

I could fix that by modifying the advised sequence from
`INSERT INTO oc_jobs(class, argument, last_run) VALUES ('OCA\Files_Trashbin\BackgroundJob\ExpireTrash', 'null', 0)`

to
`INSERT INTO `oc_jobs`(`class`, `argument`, `last_run`) VALUES ('OCA\\Files_Trashbin\\BackgroundJob\\ExpireTrash', 'null', 0)
`

However, the Trash-bin is still not cleaned   :-(

**Comment by PVince81** [comment]
Please reread https://github.com/owncloud/core/issues/22537#issuecomment-190291041

It looks like when you inserted the job manually, the backslashes got stripped.

The name is supposed to look like `OCA\Files_Trashbin\BackgroundJob\ExpireTrash`, not `OCAFiles_TrashbinBackgroundJobExpireTrash`.

I see you tried doubling the backslashes, which could fix that.
How does the "oc_jobs" table look like after that ? Is the name correct now ?

**Comment by steve-s-71** [comment]
Hi Pvince81, yes the path is correct since I doubled the backslashes. 

Other then seen when I wrote my last post - over time - I got additional entries in the log file. One was indicating there was an unsuccessful trail to clean an old file from the trash-bin.  Indeed that file was special, I could not even delete it manually.

To get rid of it, it was required to recover the file from Trash-bin to the original location, deactivate the trash-bin app, create a new folder, move the file into that folder and finally delete the whole folder with the file included.

After Trash-bin app reactivation and next cron-run, my Trash-bin got finally fully cleaned as defined in the config.php

My conclusion is the broken auto deletion was a two-fold issue: 
1. the missing path to job.php in database caused that the cleanup was not triggered
2. after that was fixed, the cleanup stuck on the file which could not get deleted

Now everything seems to work fine and I close this issue. Thanks to @nickvergessen for the useful hints.
